contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol

```diff
@@ -17,297 +17,222 @@
-            delegatedStakeByPoolId[poolId],
-            shadowRewardsInPoolByOwner[member][poolId],
-            shadowRewardsByPoolId[poolId],
-            poolBalance
-        );
+        // copy our most up-to-date cumulative rewards for last epoch, if necessary.
+        // this is necessary if the pool does not earn any rewards this epoch;
+        // if it does then this value may be overwritten when the epoch is finalized.
+        if (!_isCumulativeRewardSet(cumulativeRewardsByPoolPtr[epoch])) {
+            cumulativeRewardsByPoolPtr[epoch] = mostRecentCumulativeRewards;
+        }
     }
 
-    /// @dev A member joins a staking pool.
-    /// This function increments the shadow balance of the member, along
-    /// with the total shadow balance of the pool. This ensures that
-    /// any rewards belonging to existing members will not be diluted.
-    /// @param poolId Unique Id of pool to join.
-    /// @param member The member to join.
-    /// @param amountOfStakeToDelegate The stake to be delegated by `member` upon joining.
-    /// @param totalStakeDelegatedToPool The amount of stake currently delegated to the pool.
-    ///                                  This does not include `amountOfStakeToDelegate`.
-    function _joinStakingPool(
+    /// @dev Records a reward for delegators. This adds to the `cumulativeRewardsByPool`.
+    /// @param poolId Unique Id of pool.
+    /// @param reward to record for delegators.
+    /// @param amountOfDelegatedStake the amount of delegated stake that will split this reward.
+    /// @param epoch at which this was earned.
+    function _recordRewardForDelegators(
         bytes32 poolId,
-        address payable member,
-        uint256 amountOfStakeToDelegate,
-        uint256 totalStakeDelegatedToPool
+        uint256 reward,
+        uint256 amountOfDelegatedStake,
+        uint256 epoch
     )
         internal
     {
-        // update delegator's share of reward pool
-        uint256 poolBalance = getBalanceOfMembersInStakingPoolRewardVault(poolId);
-        uint256 buyIn = LibRewardMath._computeBuyInDenominatedInShadowAsset(
-            amountOfStakeToDelegate,
-            totalStakeDelegatedToPool,
-            shadowRewardsByPoolId[poolId],
-            poolBalance
+        // cache a storage pointer to the cumulative rewards for `poolId` indexed by epoch.
+        mapping (uint256 => IStructs.Fraction) storage cumulativeRewardsByPoolPtr = cumulativeRewardsByPool[poolId];
+
+        // fetch the last epoch at which we stored an entry for this pool;
+        // this is the most up-to-date cumulative rewards for this pool.
+        uint256 cumulativeRewardsLastStored = cumulativeRewardsByPoolLastStored[poolId];
+        IStructs.Fraction memory mostRecentCumulativeRewards = cumulativeRewardsByPoolPtr[cumulativeRewardsLastStored];
+
+        // compute new cumulative reward
+        (uint256 numerator, uint256 denominator) = LibFractions.addFractions(
```

Comment:
Would it make more sense for `addFractions` to return a fraction? It seems like we create a Fraction object a significant amount of the time after calling this.

Reply:
Mm idk, since this is a general util library I'd prefer to not force a struct on future users. We actually only convert it to the Fraction struct in this fn for storage.